src/CSET/operators/collapse.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2022 Met Office and contributors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Operators to perform various kind of collapse on either 1 or 2 dimensions.
+"""
+
+import iris
+import iris.cube
+import iris.analysis
+
+
+def collapse_1dim(
+    cube: iris.cube.Cube, coordinate: str, method: str, **kwargs
+) -> iris.cube.Cube:
+    """
+    Collapses similar (stash) fields in a cube into a cube collapsing around the specified coordinate and method.
+    This could be a (weighted) mean or percentile.
+
+    Arguments
+    ---------
+    cube: iris.cube.Cube
+         Cube to collapse and iterate over one dimension
+    coordinate: str
+         Coordinate to collapse over i.e.
+         'time', 'longitude', 'latitude','model_level_number'
+    method: str
+         Type of collapse i.e. method: 'MEAN', 'MAX', 'MIN', 'MEDIAN', 'PERCENTILE'
+         getattr creates iris.analysis.MEAN, etc
+         For PERCENTILE YAML file requires i.e.
+         method: 'PERCENTILE'
+         additional_percent: 90
+
+    Returns
+    -------
+    cube: iris.cube.Cube
+         Single variable but several methods of aggregation
+
+
+    Raises
+    ------
+    ValueError
+    If the constraint doesn't produce a single cube containing a field.
+    """
+
+    if method != "PERCENTILE":
+        collapsed_cube = cube.collapsed(coordinate, getattr(iris.analysis, method))
+    if method == "PERCENTILE":
+        for num in kwargs.values():
+            collapsed_cube = cube.collapsed(
+                coordinate, getattr(iris.analysis, method), percent=num
+            )
+
+    return collapsed_cube
+
+
+def collapse_cube_2dim(
```

Comment:
One idea I had is to make the collapse operator agnostic to the number of coordinates being collapsed over.

We could have it so the coordinate can either be a string or a list. If it is a string we collapse over that coord. If it is a list we collapse over all of the coords in that list.

That would generalise the operator so we wouldn't need a 2dim version, while also allowing more dimensions to be collapsed if ever desired. E.g. an ensemble domain mean.

Reply:
I would prefer that. The reason I initially had the two 1dim and 2dim functions to make it clear that we can also do zonal means etc. If updated in the documentation that the collapse can be done for 1 or 2 dimensions then I am happy with your suggestion.